Wake On Lan

mpotrykus opened the pull request:
Built off of the work of sixk32's PR.

Main changes:
- Separated IP Address and Port from mdnshandler
- Created further validation and dynamic handling for IP and broadcast addresses.

Notes:
- This version validates the given IP address. If it is the host name, it will retrieve the attached IP address. Tested by adding host by host name manually.
- Broadcast address is now configured dynamically.
- A lot of the work in the wake function can be broken down into smaller functions for cleaner code, but as these are single use, I decided it was unnecessary. Happy to clean things up and break it out if it helps though.

Happy to work on any changes you can see or answer any questions.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**  
 - Introduced a “Wake Host” option in the host selection interface, enabling users to quickly wake a host.  
 - Enhanced host setup now requires an IP address and port, improving connection accuracy.  
 - Added support for retrieving and displaying network details, including server MAC addresses, to improve connectivity insights.  
 - New dialog for adding hosts with structured input fields for IP address and port.  
- **Bug Fixes**  
 - Improved error messaging related to host connection failures.  
- **UI Enhancements**  
 - Updated displayed host information to reflect current hostname instead of last hostname across various settings and interfaces.  
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

TheElixZammuto commented on Pages/HostSelectorPage.xaml.cpp:
```diff
@@ -56,14 +56,14 @@ void moonlight_xbox_dx::HostSelectorPage::NewHostButton_Click(Platform::Object^
 	auto def = args->GetDeferral();
-	Concurrency::create_task([def, hostname, this, args, sender]() {
-		bool status = state->AddHost(hostname);
+	Concurrency::create_task([def, ipAddress, this, args, sender]() {
```
Replacing the bot comment here, users expect to input the port number too here and not hust the IP Address, valid input combination can be:
-192.168.1.1
- 192.168.1.1:47989
- DESKTOP-AAAA:47989 (via local hostname resolution)
- servername.mymoonlighthostingservice.com:47989 (some niche providers do expose these hosts over the internet using DNS names)

We can split IP and port directly here instead of keeping lastHostname

TheElixZammuto commented on State/ApplicationState.cpp:
```diff
@@ -80,6 +88,8 @@ Concurrency::task<void> moonlight_xbox_dx::ApplicationState::UpdateFile()
 		stateJson["alternateCombination"] = that->AlternateCombination;
 		for (auto host : that->SavedHosts) {
 			nlohmann::json hostJson;
+			hostJson["ipAddress"] = Utils::PlatformStringToStdString(host->IPAddress);
```
What happens when a user with an older version of the app opens it? 
ipAddress/port might not be populated when upgrading

TheElixZammuto reviewed (changes requested):
Some points to adjust

I've also removed the out-of-scope AI Bot comments, kept only the one I think are worth to check out

mpotrykus commented:
> Some points to adjust
> 
> I've also removed the out-of-scope AI Bot comments, kept only the one I think are worth to check out

Sounds good! I will start going through to address and test these comments this week.

mpotrykus commented:
I addressed all of the comments, but I discovered a bug on testing. You can't resolve by hostname if the machine is not connected.🤦

I don't like how messy this is getting with handling the old "lastHostname" either. I am investigating if we can do a better version with just grabbing the mac address on pairing and using WoL with that.

I'll either update this PR if I find a solution for this version, or create a new PR and branch if the mac address works better.

mpotrykus commented on State/ApplicationState.cpp:
```diff
@@ -125,4 +136,135 @@ moonlight_xbox_dx::ApplicationState^ __stateInstance;
+	}
+
+	closesocket(s);
+	WSACleanup();
```
WoL must be sent over port 7 or 9. This change broke it.

mpotrykus commented on State/ApplicationState.cpp:
```diff
@@ -125,4 +136,135 @@ moonlight_xbox_dx::ApplicationState^ __stateInstance;
+			WSACleanup();
+			return;
+		}
+	}
```
Added

TheElixZammuto commented:
Superseded by #165
